File: Rust/973.k-closest-points-to-origin.rs
/*
 * @lc app=leetcode id=973 lang=rust
 *
 * [973] K Closest Points to Origin
 */

// @lc code=start
impl Solution {
    pub fn k_closest(points: Vec<Vec<i32>>, k: i32) -> Vec<Vec<i32>> {
        let K = k as usize;
        let mut res = points.clone();
        Self::quick_select(&mut res, K);
        res = res[..K].to_vec();
        return res;
    }
    
    fn quick_select(arr: &mut Vec<Vec<i32>>, K: usize) {
        let mut lo:usize = 0;
        let mut hi:usize = arr.len() - 1;
        while(lo < hi) {
            let mut pivotIdx = Self::partition(arr, lo, hi);
            if(pivotIdx == K) {
                break;
            } else if(pivotIdx < K) {
                lo = pivotIdx + 1;
            } else {
                hi = pivotIdx - 1;
            }
        }
    }
    
    fn partition(arr: &mut Vec<Vec<i32>>, low: usize, high: usize) -> usize {
        let pivot = arr[high].clone();
        println!("Pivot: {:?}", pivot);
        let mut pivotIdx = low;
        let mut lo = low;
        let mut hi = high;
        for lo in lo..hi {
            if(Self::cmp(&arr[lo], &pivot) < 0) {
                arr.swap(pivotIdx, lo);
                pivotIdx += 1;
            }
        }
        arr.swap(pivotIdx, hi);
        pivotIdx
    }
    
    fn cmp(p1: &Vec<i32>, p2: &Vec<i32>) -> i32{
        return p1[0]*p1[0] + p1[1]*p1[1] - p2[0] * p2[0] - p2[1]*p2[1];
    }
}
// @lc code=end

File: JS/973.k-closest-points-to-origin.js
/*
 * @lc app=leetcode id=973 lang=javascript
 *
 * [973] K Closest Points to Origin
 */

// @lc code=start
/**
 * @param {number[][]} points
 * @param {number} K
 * @return {number[][]}
 */
/*
利用quick sort快排的原理，当pivotIdx等于K的时候，pivotIdx左边的都是
较小的数，右边都是较大的数。
*/
var kClosest = function(points, K) {
  if(points === null || points.length === 0) return [];
  quickSelect(points);
  return points.slice(0, K);
  
  function quickSelect(points) {
    let len = points.length;
    let lo = 0, hi = len - 1;

    while(lo < hi) {
      let pivotIdx = partition(points, lo, hi);
      if(pivotIdx === K) {
        break;
      } else if(pivotIdx < K) {
        lo = pivotIdx + 1;
      } else {
        hi = pivotIdx - 1;
      }
    }
  }
  
  function partition(arr, low, high) {
    let pivot = arr[high];
    let pivotIdx = low;
    for(let i = low; i < high; i++) {
      if(cmp(arr[i], pivot) < 0) {
        swap(arr, i, pivotIdx);
        pivotIdx++;
      }
    }
    swap(arr, pivotIdx, high);
    return pivotIdx;
  }
  
  
  function swap(arr, p, q) {
    let tmp = arr[p];
    arr[p] = arr[q];
    arr[q] = tmp;
  }
  
  function cmp(p1, p2) {
    return p1[0] * p1[0] + p1[1] * p1[1] - p2[0] * p2[0] - p2[1] * p2[1];
  }
};
// @lc code=end

